// msda-storage/esm2015/lib/msda-storage.module.js
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { MsdaStorage } from './msda-storage.service';
import * as i0 from "@angular/core";
export class MsdaStorageModule {
}
MsdaStorageModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MsdaStorageModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
MsdaStorageModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MsdaStorageModule, imports: [HttpClientModule,
        CommonModule] });
MsdaStorageModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MsdaStorageModule, providers: [MsdaStorage], imports: [[
            HttpClientModule,
            CommonModule
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.2.16", ngImport: i0, type: MsdaStorageModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [
                        HttpClientModule,
                        CommonModule
                    ],
                    exports: [],
                    providers: [MsdaStorage]
                }]
        }] });
